File: src/DesignPattern.VBNet/Sigletons/SingletonLock.vb

Namespace DesignPattern.VBNet.Sigletons
    '雙重上鎖 VB.Net
    Public Class SingletonLock
        Private Shared instance As SingletonLock
        Private Shared ReadOnly syncRoot As Object = New Object()
        Private Sub New()
        End Sub
        Public Shared Function GetInstance() As SingletonLock
            If instance Is Nothing Then
                SyncLock syncRoot
                    If instance Is Nothing Then
                        instance = New SingletonLock()
                    End If
                End SyncLock
            End If
            Return instance
        End Function
    End Class

End Namespace

File: src/DesignPattern.CSharp.Sigletons/SingletonLock.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DesignPattern.CSharp.Sigletons
{

    //雙重上鎖
    public class SingletonLock
    {
        private static SingletonLock instance;
        private static readonly object syncRoot = new object();
        private SingletonLock() { }
        public static SingletonLock GetInstance()
        {
            if (instance == null) //確保當前沒有這個物件存在
            {
                lock (syncRoot)//鎖住當前狀態
                {
                    if (instance == null)
                        instance = new SingletonLock();
                }
            }
            return instance;
        }
    }
}
